src/main/scala/chiseltest/internal/SingleThreadBackend.scala

```diff
@@ -0,0 +1,45 @@
+package chiseltest.internal
+
+import chisel3.{Data, Module}
+import chiseltest.coverage.TestCoverage
+import chiseltest.simulator.SimulatorContext
+import chiseltest.testableClock
+import firrtl.AnnotationSeq
+
+class SingleThreadBackend[T <: Module](
+  val dut:             T,
+  val dataNames:       Map[Data, String],
+  val tester:          SimulatorContext,
+  coverageAnnotations: AnnotationSeq)
+    extends GenericBackend[T]
+    with BackendInstance[T] {
+  override def expectBits(
+    signal:  Data,
+    value:   BigInt,
+    message: Option[String],
+    decode:  Option[BigInt => String],
+    stale:   Boolean
+  ): Unit = {
+    super.expectBits(signal, value, message, decode, stale)
+    Context().env.checkpoint()
```

Comment:
checkpoint at once, this is a little different then `ThreadedBackend`, but I think this is an improvement.

Reply:
Could you write a test that makes sure that the errors reported are the same no matter whether threading is enabled or not?

Reply:
This won’t affect threading code,, this is just for single thead api. and it’s reasonable to checkpoint at schedule for threading api

Reply:
Basically what I would like to ensure is that if you have the exact same test run with the two different backends, the errors reported should be 100% the same. Switching the backends should not affect any behavior besides the fact that `fork` won't work in a non-threaded context.

Reply:
you mean "two different backends" is for both threading and non-threading?
I prefer achieving this in the future, after refactoring entire chiseltest into the Stage/Phases as the last time I did.

Reply:
> the errors reported should be 100% the same

I think this is for different "simulator backends" like Verilator, VCS, NCSIM etc. But for different API, I think this restriction can be relaxed, in this case, the only difference is whether report asserting failed in this cycle or next cycle.
`ThreadedBackend` need to report in the next cycle to gather all exception among the thread.
However `SingleThreadAPI` only contains one thread, and can throw exception at his cycle.

Reply:
> `ThreadedBackend` need to report in the next cycle to gather all exception among the thread.
> However `SingleThreadAPI` only contains one thread, and can throw exception at his cycle.

In order for the threading option to only be an optimization hint, both SingleThreaded and ThreadedBackend need to behave 100% the same on tests that do not use fork.

Reply:
> In order for the threading option to only be an optimization hint, both SingleThreaded and ThreadedBackend need to behave 100% the same on tests that do not use fork.

I see, so you mean, for entire tests, we should ran both SingleThreaded and Threaded successfully?

Reply:
It is enough if you craft 2-3 example tests that make sure both threaded and single threaded behave the same.

Reply:
Got it :)